docs_src/microservices/security/Ch-Docker-Swarm-SecuringDeviceServices.md

```diff
@@ -0,0 +1,229 @@
+
+```sh
+sudo docker swarm join --token <token> <manager ip address>:2377
+```
+
+Next, I will walk-through the changes we made to the docker-stack.yml file to convert the edgex compose file into a docker swarm stack file.
+
+### Setting up the docker-stack-edgex.yml file
+
+All of the following changes are already done in the examples repo.  I will just outline the necessary changes to go from a compose file to stack file.
+
+First, remove 'restart' command from compose file; 'restart' is not a valid command in docker swarm.
+
+Next, we define constraints that the edgex core services must not run on the worker node add this section of yml to the 'docker-stack-edgex.yml'.  We will do the inverse for the device-service to ensure it does run on the worker node thus ensuring it uses the overlay network to communicate with the other services.  Note that this is already done in the example directory.
+
+```yaml
+    deploy: 
+      placement: 
+        constraints: 
+          - node.hostname != worker-node
+```
+Here is the inverse of the previous yml block. This gets added to the device services in the stack file. 
+
+```yaml
+    deploy: 
+      placement: 
+        constraints: 
+          - node.hostname == worker-node
+```
+
+These work because we set the 'hostname = worker-node' in the Vagrantfile.
+
+### Adding Host Mounted Volumes
+
+In docker swarm bind mount volumes need to be explicitly defined in the stack yml file.  Below are the first three bind mount volume definitions these directories must be created on the host before the stack file can be ran.
+
+Note that, this is only an example in a production deployment you would want to use a network filesystem or create the shared volumes between containers.
+
+```yaml
+  secrets-volume:
+    driver: local
+    driver_opts:
+      o: bind
+      type: none
+      device: /tmp/edgex/secrets/
+  secrets-ca-volume:
+    driver: local
+    driver_opts:
+      o: bind
+      type: none
+      device: /tmp/edgex/secrets/ca/
+  edgex-consul:
+    driver: local
+    driver_opts:
+      o: bind
+      type: none
+      device: /tmp/edgex/secrets/edgex-consul/
+...
+```
+
+The full docker-stack file is included here:
+
+[docker-stack-edgex.yml file](https://github.com/edgexfoundry/edgex-examples/security/remote_devices/docker-swarm/docker-stack-edgex.yml)
+
+### Other changes in docker-stack-edgex.yml file 
+
+Another change we had to make in the docker-stack-edgex.yml file is to disable `IPC_LOCK` because the `cap_add` flag in vault's configuration is not supported in docker swarm.  To do this we add `SKIP_SETCAP=true` and `disable_mlock = "true"` to vault in the stack file. 
+
+```yaml
+  vault:
+    image: vault:1.3.1
+    hostname: edgex-vault
+    networks:
+      edgex-network:
+        aliases:
+            - edgex-vault
+    ports:
+      - "8200:8200"
```

Comment:
Still need to pull in the changes we made in the other file and make them consistent.